packages/stream-model-handler/src/__tests__/model-handler.test.ts

```diff
@@ -110,7 +110,7 @@ const CONTENT_WITH_INVALID_VIEWS: ModelDefinition = {
 
 const CONTENT_WITH_INVALID_SCHEMA = {
   name: 'MyModel',
-  accountRelation: ModelAccountRelation.LIST,
+  accountRelation: { type: 'list' },
```

Comment:
Technicality here. While I understand you love string literals, they will make maintenance harder. Could you make `list` or `single` an enum back, please?

Reply:
The goal is that the `ModelAccountRelation` itself is an object type, with its `type` as discriminant so it would be completely restricted by the TS compiler. If we need to use an enum, are you suggesting we have both `ModelAccountRelation` and `ModelAccountRelationType` please?

Also, do you have examples of how enums make maintenance easier please? In my experience it's the opposite, enums are a pain to reuse in "type only" modules. For example to use them in ComposeDB packages, I need to add a runtime dependency (can't be a dev dependency) on `@ceramicnetwork/stream-model` simply to import the enum, which makes the dependency tree heavier only for type constraints.

Reply:
> The goal is that the `ModelAccountRelation` itself is an object type, with its `type` as discriminant so it would be completely restricted by the TS compiler. If we need to use an enum, are you suggesting we have both `ModelAccountRelation` and `ModelAccountRelationType` please?

TS Compiler is smart enough to treat union of interfaces with distinct `type` field as a discriminated union, exactly the same as with the strings.

```
enum ModelAccountRelationType {
  LIST = 'list',
  SINGLE = 'single'
}
type ListRelation = {type: ModelAccountRelationType.LIST}
const ListRelation: ListRelation = {type: ModelAccountRelationType.LIST}
type SingleRelation = {type: ModelAccountRelationType.SINGLE}
const SingleRelation: SingleRelation = {type: ModelAccountRelationType.SINGLE}
type AccountRelation = ListRelation | SingleRelation

function foo(listRel: ListRelation) {
}
foo(ListRelation) // error
```

Also, `switch` works exhaustive too.

> Also, do you have examples of how enums make maintenance easier please? In my experience it's the opposite, enums are a pain to reuse in "type only" modules. For example to use them in ComposeDB packages, I need to add a runtime dependency (can't be a dev dependency) on `@ceramicnetwork/stream-model` simply to import the enum, which makes the dependency tree heavier only for type constraints.

I am curious, why is it a big trouble? Is it one of the biggest dependencies?

Examples of how enums make life easier:
1. Initially Ceramic had 3 pubsub messages, typed through enum. Message processing used exhaustive switch statement. Then we decided to add 4th. Now the compiler suggested where to add code to process the 4th type. Less error prone, more clear code.
2. Once we added one more SyncOption to Ceramic: `NEVER_SYNC`. It is pretty easy to find code probably affected by this addition. In case of string literals, it would be just plain search. For example, here, if we change something related to `list` type, we would have to sieve through all the `list` words to find relevant places.

Reply:
> > The goal is that the `ModelAccountRelation` itself is an object type, with its `type` as discriminant so it would be completely restricted by the TS compiler. If we need to use an enum, are you suggesting we have both `ModelAccountRelation` and `ModelAccountRelationType` please?
> 
> TS Compiler is smart enough to treat union of interfaces with distinct `type` field as a discriminated union, exactly the same as with the strings.
> 
> ```
> enum ModelAccountRelationType {
>   LIST = 'list',
>   SINGLE = 'single'
> }
> type ListRelation = {type: ModelAccountRelationType.LIST}
> const ListRelation: ListRelation = {type: ModelAccountRelationType.LIST}
> type SingleRelation = {type: ModelAccountRelationType.SINGLE}
> const SingleRelation: SingleRelation = {type: ModelAccountRelationType.SINGLE}
> type AccountRelation = ListRelation | SingleRelation
> 
> function foo(listRel: ListRelation) {
> }
> foo(ListRelation) // error
> ```
> 
> Also, `switch` works exhaustive too.

Yes I agree, that's where I don't see any advantage of using enums instead of regular type objects, considering TS will provide the same safety either way?

> > Also, do you have examples of how enums make maintenance easier please? In my experience it's the opposite, enums are a pain to reuse in "type only" modules. For example to use them in ComposeDB packages, I need to add a runtime dependency (can't be a dev dependency) on `@ceramicnetwork/stream-model` simply to import the enum, which makes the dependency tree heavier only for type constraints.
> 
> I am curious, why is it a big trouble? Is it one of the biggest dependencies?

Not the biggest, but to me the question is more the other way around: why add runtime dependencies that are not strictly needed?

> Examples of how enums make life easier:
> 
>     1. Initially Ceramic had 3 pubsub messages, typed through enum. Message processing used exhaustive switch statement. Then we decided to add 4th. Now the compiler suggested where to add code to process the 4th type. Less error prone, more clear code.
> 
>     2. Once we added one more SyncOption to Ceramic: `NEVER_SYNC`. It is pretty easy to find code probably affected by this addition. In case of string literals, it would be just plain search. For example, here, if we change something related to `list` type, we would have to sieve through all the `list` words to find relevant places.

Thanks, that's good examples. Indeed it makes it easier for search in the code-base, but I've been through similar cases refactoring the ComposeDB logic to add the extra relations and the experience is pretty much the same, the TS compiler will error if there are missing cases due to the added types, I don't think enums would have made a difference here?

To me really the biggest concern is the need for extra runtime dependency on this package. It's essentially "leaking" the type abstraction to any downstream app/package, leading to more risks of duplicate dependencies, mismatches, etc.
We also haven't made any effort to address overall bundle size recently but I think it will be something to put more effort on over time, so the more we can limit the number of runtime dependencies we need the better.

Reply:
Oh, okay. Let it be this way with the literals.